libguard: Added persistentTypeOnly flag in the getAll interface

RameshIyyar opened the pull request:
- Currently, the getAll interface returning all types (persistent and
  ephemeral) of guard records but, the ephemeral type records are
  used for internal purpose i.e, between the BMC and Host and those type
  of guard records not intended to expose to the end-user.

- So, the application (that used to list records to the end-user) is
  adding the filter to ignore the ephemeral type records so simplified
  the getAll interface to return only persistent type records if requested
  else return all types of records.

Tested:

- Verified by creating and listing the guard records.

RameshIyyar committed:
libguard: Added persistentTypeOnly flag in the getAll interface

- Currently, the getAll interface returning all types (persistent and
  ephemeral) of guard records but, the ephemeral type records are
  used for internal purpose i.e, between the BMC and Host and those type
  of guard records not intended to expose to the end-user.

- So, the application (that used to list records to the end-user) is
  adding the filter to ignore the ephemeral type records so simplified
  the getAll interface to return only persistent type records if requested
  else return all types of records.

Tested:

- Verified by creating and listing the guard records.

Signed-off-by: Ramesh Iyyar <rameshi1@in.ibm.com>
Change-Id: I42e8b41327f0b3d19d3dfba56b6c179eae669433

devenrao commented on libguard/guard_interface.cpp:
```diff
@@ -260,14 +260,19 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType,
     GuardFile file(guardFilePath);
     for_each_guard(file, pos, curRecord)
     {
+        if (persistentTypeOnly && (curRecord.errType == GARD_Reconfig ||
```
you want to fetch if persistentTypeOnly but the condition above looks like if flag set ignore

devenrao replied:
you did not capture in comments about checking for reconfig and sticky deconfig? what is that for

devenrao replied:
How do we determine "the ephemeral type records are  used for internal purpose"

devenrao replied:
is this condition saying is persistent and ephemeral ? with those two checks?

RameshIyyar commented on libguard/guard_interface.cpp:
```diff
@@ -260,14 +260,19 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType,
     GuardFile file(guardFilePath);
     for_each_guard(file, pos, curRecord)
     {
+        if (persistentTypeOnly && (curRecord.errType == GARD_Reconfig ||
```
GARD_Reconfig and GARD_Sticky_deconfig are ephemeral type records. If the retrieved record type is ephemeral and the caller intended to get persistentTypeOnly record means, just skipping and moving to the next record.

is it confusing...?

ojayanth commented on libguard/guard_interface.cpp:
```diff
@@ -260,14 +260,19 @@ GuardRecord create(const EntityPath& entityPath, uint32_t eId, uint8_t eType,
 }
 
-GuardRecords getAll()
+GuardRecords getAll(bool persistentTypeOnly)
```
The direction used here is get all the available gurad records using this api, It is better to implment lib users to filterout the require guard records here. like guard tool  implemented  https://github.com/open-power/guard/blob/main/guard.cpp#L27  .

RameshIyyar replied:
Still, this API supports to return of all records i.e by default it returns all records.

IMO, we are forcing the caller to do the filter that he doesn't want to do forever as per the direction and one more issue that I am seeing is, whenever introducing a new ephemeral type of record, we are expecting to change in the caller side too so, this change allows us to keep all in the one place and also give the flexibility to get all records too.

are you seeing any problem with this proposal?

ojayanth replied:
The commit i mentioned earlier is also pushed by you and i remember we were discussion same  and decided to go with the direction i mentioned earlier. .Better be consistent in guard tool inxcase you want to go with api change unless ther is strong reason to do that.

RameshIyyar replied:
ah ok, I did not push that code, it is already there before I start to contribute to the guard. I have extended that for [a new type](https://github.com/open-power/guard/pull/8).

Now, I am trying to fix something else but, due to this filter logic at the caller end, it forced me to do extra computation that's not needed by the caller forever. 

I can change the guard tool code in the same way.

RameshIyyar committed:
guard: Updated libguard getAll interface as per signature

- The libguard getAll interface signature is changed to get only
  persistent type records.

- So, updated the getAll interface usage in the guard tool too
  since it required the persistent type records and also removed
  filter to ignore ephemeral type records.

Tested:

- Verified by the `guard -l`.

Signed-off-by: Ramesh Iyyar <rameshi1@in.ibm.com>
Change-Id: I7f771318672fc82a2b43f6c194890020418413ac
